src/Random.mo

```diff
@@ -0,0 +1,303 @@
+  /// thus the uniformity of the distributions is *not* guaranteed.
+  public class Cyclic(entropy : Blob) {
+    let it : { next : () -> ?Word8 } = {
+      var inner = entropy.bytes();
+      next = func () : ?Word8 =
+      switch (inner.next()) {
+        case null { inner := entropy.bytes(); inner.next() };
+        case w w
+      }
+    };
+
+    /// Distributes outcomes in the numeric range [0 .. 255].
+    public func byte() : Nat8 {
+      switch (it.next()) {
+        case (?w) Prim.word8ToNat8 w;
+        case _ P.unreachable();
+      }
+    };
+
+    /// Bool iterator splitting up a byte of entropy into 8 bits
+    let bit : { next : () -> ?Bool } = {
+      var mask = 0x80 : Word8;
+      var byte = 0x00 : Word8;
+      next = func () : ?Bool {
+        if (0 : Word8 == mask) {
+          switch (it.next()) {
+            case null P.unreachable();
+            case (?w) {
+              byte := w;
+              mask := 0x40;
+              ?(0 : Word8 != byte & (0x80 : Word8))
+            }
+          }
+        } else {
+          let m = mask;
+          mask >>= (1 : Word8);
+          ?(0 : Word8 != byte & m)
+        }
+      }
+    };
+
+    /// Simulates a coin toss.
+    public func coin() : Bool {
+      switch (bit.next()) {
+        case (?w) w;
+        case _ P.unreachable();
+      }
+    };
+
+    /// Distributes outcomes in the numeric range [0 .. 2^n - 1].
+    public func range(p : Nat8) : Nat {
+      var pp = p;
+      var acc : Nat = 0;
+      for (i in it) {
+        if (8 : Nat8 <= pp)
+        { acc := acc * 256 + Prim.word8ToNat(i) }
+        else if (0 : Nat8 == pp)
+        { return acc }
+        else {
+          acc *= Prim.word8ToNat(1 << Prim.nat8ToWord8 pp);
+          let mask : Word8 = -1 >> Prim.nat8ToWord8(8 - pp);
+          return acc + Prim.word8ToNat(i & mask)
+        };
+        pp -= 8
+      };
+      P.unreachable()
+    };
+
+    /// Counts the number of heads in `n` coin tosses.
+    public func binomial(n : Nat8) : Nat8 {
+      var nn = n;
+      var acc : Word8 = 0;
+      for (i in it) {
+        if (8 : Nat8 <= nn)
+        { acc += Prim.popcntWord8(i) }
+        else if (0 : Nat8 == nn)
+        { return Prim.word8ToNat8 acc }
+        else {
+          let mask : Word8 = -1 << Prim.nat8ToWord8(8 - nn);
+          let residue = Prim.popcntWord8(i & mask);
+          return Prim.word8ToNat8(acc + residue)
+        };
+        nn -= 8
+      };
+      P.unreachable()
+    }
+  };
+
+  let raw_rand = (actor "aaaaa-aa" : actor { raw_rand : () -> async Blob }).raw_rand;
+
+  /// Uniformly distributes outcomes in the numeric range [0 .. 255].
+  public func byte() : async Nat8 {
+    let bytes = await raw_rand();
+    byteFrom bytes
+  };
+
+  /// Distributes outcomes in the numeric range [0 .. 255].
+  /// Seed blob must contain at least a byte.
+  public func byteFrom(seed : Blob) : Nat8 {
+    switch (seed.bytes().next()) {
```

Comment:
Might as well use `Option.unwrap()` for these.

Reply:
@kritzcreek, plan still is to remove Option.unwrap once we have option monad syntax.

Reply:
But only at the cost of pulling in another module. Not sure how the recommendations go. I'd be more willing if we had an inlining story that didn't add extra calls. Leaving for now.